### ISSUE TITLE
[MNG-5977] Improve output readability of our MavenTransferListener implementations

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7727]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** opened **[MNG-5977](https://issues.apache.org/jira/browse/MNG-5977?redirect=false)** and commented

The current output of Downloading/Downladed/Uploading/Uploaded transfer notification has some flaws:

1. It does not scale numbers between 1 and 1000 with appropriate units
2. It should use correct size (`kB`, `MB`, `GB` and time units (`s`) but doesn't. (see https://en.wikipedia.org/wiki/Binary_prefix and https://en.wikipedia.org/wiki/Metric_prefix)
3. When Aether downloads in parallel (which applies for non-POM files) the progress interleaves due to race conditions to `System.out` and you do not know to which resource a progress belongs to.

Let's use an improved version of MPIR `DependenciesRenderer`'s [`FileDecimalFormat`](https://github.com/apache/maven-plugins/blob/trunk/maven-project-info-reports-plugin/src/main/java/org/apache/maven/report/projectinfo/dependencies/renderer/DependenciesRenderer.java#L1583) for it.

concrete examples:
before

```
191/191 KB   27/48 KB   48/119 KB   80/87 KB   13/13 KB   
```

after:

```
Progress (4): 500/800 B | 40/45 kB | 193 kB/315 kB | 1.3/9.0 MB | 12/30 MB
```

if total size is unavailable or the file has already been downloaded but not removed from the list, the output will be:

```
Progress (4): 800 B | 40/45 kB | 193 kB | 9.0 MB | 12 MB
```

or in debug mode:

```
Progress (5): xml-apis-1.3.04.jar (<progress>) | maven-shared-utils-0.6.jar (<progress>) | xercesImpl-2.9.1.jar (<progress>) | commons-digester-1.6.jar (<progress>) | maven-reporting-impl-2.3.jar (<progress>)
```

If the scale is between 1 and 10, one decimal place will be printed out. If it is between 10 and 1000+, it will be an integer.

---

**Affects:** 3.3.9

**Attachments:**
- [current.zip](https://issues.apache.org/jira/secure/attachment/12787978/current.zip) (_77.56 kB_)
- [improved.zip](https://issues.apache.org/jira/secure/attachment/12789760/improved.zip) (_77.52 kB_)
- [improved-debug.zip](https://issues.apache.org/jira/secure/attachment/12789761/improved-debug.zip) (_516.16 kB_)

**Remote Links:**
- [Page
](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=67636510)

**Backported to:** [3.5.0-alpha-1](https://github.com/apache/maven/milestone/92?closed=1)

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

can you give some concrete examples of the issue and the proposed solution output, please?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I will provide you some today. I have already some code locally on my machine which looks good. I am still trying to solve the parallel write interleave to `System.out`. Can't figure out what I need to synchronize, `out` of `lineLength`.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Salut Hervé,

I have made some progress on the issue and have produced two versions available from: http://home.apache.org/~michaelo/MNG-5977/

1. `apache-maven-3.4.0-MNG-5977-SNAPSHOT-bin.tar.gz`: Nicely prints the progress. No race conditions, no syncs where necessary. I have already cleanup superfluous padding and `CRs`. All looks fine.
2. `apache-maven-3.4.0-MNG-5977-w-filename-SNAPSHOT-bin.tar.gz`: Extends the previous implementation by adding the filename to the progress meter. Very nice to read the progress now. The downside is that all methods writing to `PrintStream` require synchronization otherwise the race conditions will garble the output on long progress lines. (This can be improved by using a `ReentrantLock` around `lastLength` and `out` access only.) There is one glitch I do not understand and don't know how to fix it yet. Under high concurrency, the progress is written but not overwritten by the next progress output or an upcoming "Downloaded:" line. The output written by `tee` is fine. This fails in `cmd` and Git bash. This may be a race condition within the Aether threads (unlikely hopefully) or a flaw in Windows or some important flush I forgot. Anyway, I would appreciate if you could take a look at my patches, maybe you will spot the issue. The second patch requires the first one first to be applied.

In any case, I'd like to prepend the progress meter with something like "In Progress: " or "In Transfer: ". Makes is better readable.
Tested under Windows 10 only, Linux and FreeBSD is pending.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Update on the `synchronized` issue: surprisingly, all write operations of `PrintStream` are synchronized in the OpenJDK. This would mean that I could remove the keywords and synchronize only where `lastLength` is accessed (`ro` or `rw`).

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Adde commit [5efbfd780e5888febfc644582d201f0145246b90](https://git-wip-us.apache.org/repos/asf?p=maven.git;a=commit;h=5efbfd780e5888febfc644582d201f0145246b90) with slight improvements.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Issue needs refinement in code and description.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Changes in detail:

* Introduced a `FileSizeFormat` which flexibly scales the input to the nearest unit (B, kB, MB, GB)
  from 1 to 10 with one decimal place, from 10 with whole numbers. All units cleanly scale with SI prefixes.
  Additionally, a progress format is also available which scales progressed size and total size
  with the unit of the total size. An unknown total size is handled too. If progressed size and total
  size are equal (download has finished), numbers collapse. The format class has full test coverage.
  All transfer listeners use this format equally.
* Synchronize all calls of the console transfer listener to avoid race conditions for the `PrintStream`
  which happens for longer lines.
* Print overriding blank lines when necessary only.
* Prepend the progress with `Progress (n): ` to denote that first progress is going on and second how many
  transfers are in the pipeline.
* Separate parallel progresses with ` | `.
* Replaced `ConcurrentHashMap` with a synchronized `LinkedHashMap`. This retains the insertion
  order of the transfers which makes the progress meter look like a parallel progress queue. No disturbing
  jumps between progress changes anymore.
* If running in debug mode, the filename is printed additionally.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Here are two log files containing output after applied improvements.

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

Maven 3.4.0 has been dropped. See [this thread](https://mail-archives.apache.org/mod_mbox/maven-dev/201701.mbox/%3CCA%2BnPnMx-e7kGYy3Hp87v8hLGdhp1q%3DtKLx_6QuZ4kGUqHEBGcw%40mail.gmail.com%3E) for more details.

This issue will need to be re-scheduled for a Maven release in the (hopefully near) future.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [149cce7a867956efeaf72d527f61297bf2471b1e](https://git-wip-us.apache.org/repos/asf?p=maven.git;a=commit;h=149cce7a867956efeaf72d527f61297bf2471b1e).
